### PR TITLE
[13.0][MIG] sale_order_invoicing_finished_task: Migration to 13.0

### PR DESCRIPTION
**Pull request by joao-p-marques**
Supersedes https://github.com/OCA/sale-workflow/pull/1272

Taking from the previous work:
* Applied @carlosdauden's suggestion
* Split pre-commit changes into a separate commit
* Squashed administrative commits
* Fixed tests

@Tecnativa
TT27067

@carlosdauden @Tardo can you review?

### CODE REVIEW

File: sale_order_invoicing_finished_task/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# Copyright 2017-2018 Carlos Dauden <carlos.dauden@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/__manifest__.py]:
```suggestion
# Copyright 2017 Tecnativa - Sergio Teruel
# Copyright 2017-2018 Tecnativa - Carlos Dauden
```

File: sale_order_invoicing_finished_task/i18n/es.po [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#, python-format
+msgid ""
+"You cannot create/modify a task related with a invoiced, done or cancel sale "
+"order line "
+msgstr ""
+"No puede crear/modificar una tarea relacionada con una línea de pedido "
+"facturada, realizada o cancelada"
+
+#~ msgid "Quotation"
+#~ msgstr "Presupuesto"
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/i18n/es.po:86]:
```suggestion
```

File: sale_order_invoicing_finished_task/models/product.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# Copyright 2017 Carlos Dauden <carlos.dauden@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/models/product.py]:
```suggestion
# Copyright 2017 Tecnativa - Sergio Teruel
# Copyright 2017 Tecnativa - Carlos Dauden
```

File: sale_order_invoicing_finished_task/models/project.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2017 Camptocamp SA
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# Copyright 2017 Carlos Dauden <carlos.dauden@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/models/project.py]:
```suggestion
# Copyright 2017 Tecnativa - Sergio Teruel
# Copyright 2017 Tecnativa - Carlos Dauden
```

File: sale_order_invoicing_finished_task/models/project.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    def toggle_invoiceable(self):
+        self._check_sale_line_state()
+        for task in self:
+            task.invoiceable = not task.invoiceable
+
+    def write(self, vals):
+        if "sale_line_id" in vals:
+            self._check_sale_line_state(vals["sale_line_id"])
+        res = super(ProjectTask, self).write(vals)
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/models/project.py]:
```suggestion
        res = super().write(vals)
```

File: sale_order_invoicing_finished_task/models/project.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        if "invoiceable" in vals:
+            self.mapped("sale_line_id")._compute_qty_delivered()
+        return res
+
+    @api.model
+    def create(self, vals):
+        if "sale_line_id" in vals:
+            self._check_sale_line_state(vals["sale_line_id"])
+        return super(ProjectTask, self).create(vals)
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/models/project.py]:
```suggestion
        return super().create(vals)
```

File: sale_order_invoicing_finished_task/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# Copyright 2017 Carlos Dauden <carlos.dauden@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/models/sale_order.py]:
```suggestion
# Copyright 2017 Tecnativa - Sergio Teruel
# Copyright 2017 Tecnativa - Carlos Dauden
```

File: sale_order_invoicing_finished_task/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.depends(
+        "state", "order_line.invoice_status", "order_line.task_ids.invoiceable"
+    )
+    def _get_invoiced(self):
+        super(SaleOrder, self)._get_invoiced()
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/models/sale_order.py]:
```suggestion
        super()._get_invoiced()
```

File: sale_order_invoicing_finished_task/tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/__init__.py]:
```suggestion
# Copyright 2017 Tecnativa - Sergio Teruel
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
# Copyright 2017 Tecnativa - Sergio Teruel
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        cls.partner = cls.env["res.partner"].create(
+            {"name": "Customer - test", "customer_rank": True}
+        )
+        cls.project = cls.env["project.project"].create({"name": "Some test project"})
+        cls.stage_new = cls.env["project.task.type"].create(cls._prepare_stage_vals())
+        cls.stage_invoiceable = cls.env["project.task.type"].create(
+            cls._prepare_stage_vals(invoiceable_stage=True)
+        )
+        cls.uom_unit = cls.env.ref("uom.product_uom_unit")
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        product_delivery_vals.update(
+            {
+                "name": "Product - Service - Policy delivery - Test",
+                "service_policy": "delivered_timesheet",
+                "invoice_policy": "delivery",
+                "service_type": "timesheet",
+            }
+        )
+        cls.product_policy_delivery = cls.Product.create(product_delivery_vals)
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+            "product_uom_id": task.sale_line_id.product_uom.id,
+            "user_id": self.manager.id,
+            "task_id": task.id,
+        }
+
+    def test_invoice_status(self):
+        self.sale_order.action_confirm()
+        self.assertEqual(self.sale_order.invoice_status, "no")
+        task = self.sale_order.order_line.task_ids
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        # Add a timesheet line
+        self.env["account.analytic.line"].create(
+            self._prepare_timesheet_vals(task, 5.0)
+        )
+        self.assertEqual(self.sale_order.invoice_status, "no")
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+        # Set task in invoiceable stage
+        task.stage_id = self.stage_invoiceable.id
+        task._onchange_stage_id()
+        self.assertEqual(self.sale_order.invoice_status, "to invoice")
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+        # Click on toggle_invoiceable method
+        task.toggle_invoiceable()
+        self.assertEqual(self.sale_order.invoice_status, "no")
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+        task.toggle_invoiceable()
+        self.assertEqual(self.sale_order.invoice_status, "to invoice")
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+        # Make the invoice
+        self.sale_order._create_invoices()
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+        # Click on toggle_invoiceable method after the so is invoiced
+        with self.assertRaises(ValidationError):
+            task.toggle_invoiceable()
+
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/tests/test_sale_order_invoicing_finished_task.py]:
```suggestion
```

File: sale_order_invoicing_finished_task/views/product_view.xml [hunk 20]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/views/product_view.xml]:
```suggestion
<!-- Copyright 2017 Tecnativa - Sergio Teruel
```

File: sale_order_invoicing_finished_task/views/project_view.xml [hunk 21]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
```
Comment (victoralmau) [inline sale_order_invoicing_finished_task/views/project_view.xml]:
```suggestion
<!-- Copyright 2017 Tecnativa - Sergio Teruel
```

**Review by victoralmau — CHANGES_REQUESTED** [review]
Good job, only some little changes to improve code (remove emails, optimize super and remove empty lines).

File: sale_order_invoicing_finished_task/i18n/es.po (hunk 2 above)
Comment (joao-p-marques) [reply sale_order_invoicing_finished_task/i18n/es.po:86]: I think it is better to let @oca-transbot update this files and translate them later for consistency.

**Comment by joao-p-marques** [comment]
@victoralmau changes made, apart from my comment above. Thanks!

**Review by victoralmau — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**